service/history/ndc/history_replicator.go

```diff
@@ -227,6 +229,193 @@ func (r *HistoryReplicatorImpl) ApplyEvents(
+	ctx context.Context,
+	request *shard.BackfillHistoryEventsRequest,
+) error {
+	task, err := newReplicationTaskFromBatch(
+		r.clusterMetadata,
+		r.logger,
+		request.WorkflowKey,
+		request.BaseExecutionInfo,
+		request.VersionHistoryItems,
+		request.Events,
+		request.NewEvents,
+		request.NewRunID,
+		request.VersionedHistory,
+	)
+	if err != nil {
+		return err
+	}
+
+	return r.doApplyBackfillEvents(ctx, task, r.applyBackfillEvents)
+}
+
+func (r *HistoryReplicatorImpl) doApplyBackfillEvents(
+	ctx context.Context,
+	task replicationTask,
+	action func(context.Context, workflow.MutableState, workflow.Context, wcache.ReleaseCacheFunc, replicationTask) error,
+) (retError error) {
+	wfContext, releaseFn, err := r.workflowCache.GetOrCreateWorkflowExecution(
+		ctx,
+		r.shardContext,
+		task.getNamespaceID(),
+		task.getExecution(),
+		locks.PriorityHigh,
+	)
+	if err != nil {
+		return err
+	}
+	defer func() {
+		if rec := recover(); rec != nil {
+			releaseFn(errPanic)
+			panic(rec)
+		}
+		releaseFn(retError)
+	}()
+
+	mutableState, err := wfContext.LoadMutableState(ctx, r.shardContext)
+	switch err.(type) {
+	case nil:
+		return action(ctx, mutableState, wfContext, releaseFn, task)
+	case *serviceerror.NotFound:
+		return serviceerrors.NewSyncState(
+			mutableStateMissingMessage,
+			task.getNamespaceID().String(),
+			task.getWorkflowID(),
+			task.getRunID(),
+			task.getVersionedTransition(),
+		)
+	default:
+		return err
+	}
+}
+
+func (r *HistoryReplicatorImpl) applyBackfillEvents(
+	ctx context.Context,
+	mutableState workflow.MutableState,
+	wfContext workflow.Context,
+	releaseFn wcache.ReleaseCacheFunc,
+	task replicationTask,
+) (retError error) {
+
+	versionedTransition := task.getVersionedTransition()
+	if versionedTransition == nil {
+		return errors.New("versioned transition is required for backfill task")
+	}
+
+	if task.getFirstEvent().GetEventType() == enumspb.EVENT_TYPE_WORKFLOW_EXECUTION_STARTED {
+		return errors.New("workflow execution started event is not expected in backfill task")
+	}
+
+	transitionHistory := mutableState.GetExecutionInfo().GetTransitionHistory()
+	err := workflow.TransitionHistoryStalenessCheck(transitionHistory, versionedTransition)
+	if err != nil {
+		return serviceerrors.NewSyncState(
+			mutableStateMissingMessage,
+			task.getNamespaceID().String(),
+			task.getWorkflowID(),
+			task.getRunID(),
+			task.getVersionedTransition(),
+		)
+	}
+
+	mutableState, _, err = r.mutableStateMapper.FlushBufferEvents(ctx, wfContext, mutableState, task)
+	if err != nil {
+		return err
+	}
+
+	mutableState, prepareHistoryBranchOut, err := r.mutableStateMapper.GetOrCreateHistoryBranch(ctx, wfContext, mutableState, task)
+	if err != nil {
+		return err
+	} else if !prepareHistoryBranchOut.DoContinue {
+		metrics.DuplicateReplicationEventsCounter.With(r.metricsHandler).Record(
+			1,
+			metrics.OperationTag(metrics.BackfillHistoryEventsTaskScope))
+		return nil
```

Comment:
hmm do we still need to create the newRun, if exists, in this case?

Reply:
From source side, we merged new run with current run events to make it one transaction. Hence if history events are duplicated, we can assume the new run is duplicated as well. Otherwise the similar process in doApplyEvents also need to update.

Reply:
oh I guess that's why we create new run first. otherwise, event can be backfilled first, and logic fail on creating new run. Then upon retry the logic will just drop the task?

Reply:
I think that would be the case.